dbms/src/Flash/Executor/DataStreamExecutor.cpp

```diff
@@ -65,11 +69,23 @@ String DataStreamExecutor::toString() const
 
 int DataStreamExecutor::estimateNewThreadCount()
 {
-    return data_stream->estimateNewThreadCount();
+    return estimate_thread_cnt;
 }
 
 RU DataStreamExecutor::collectRequestUnit()
 {
-    return toRU(data_stream->estimateCPUTimeNs());
+    auto origin_cpu_time_ns = data_stream->estimateCPUTimeNs();
+    UInt64 total_thread_cnt = thread_cnt_before_execute + estimate_thread_cnt;
+    size_t physical_cpu_cores = getNumberOfPhysicalCPUCores();
+    if (origin_cpu_time_ns <= 0 || total_thread_cnt <= physical_cpu_cores)
+        return toRU(origin_cpu_time_ns);
+
+    // When the number of threads is greater than the number of cpu cores,
+    // BlockInputStream's estimated cpu time will be much greater than the actual value.
+    double per_thread_cpu_time_ns = static_cast<double>(origin_cpu_time_ns) / estimate_thread_cnt;
+    // It can be assumed that per_core_cpu_time = per_thread_cpu_time_ns * (total_thread_cnt / physical_cpu_cores).
+    UInt64 per_core_cpu_time_ns = ceil(per_thread_cpu_time_ns / total_thread_cnt) * physical_cpu_cores;
```

Comment:
It seems not the same for comment and code here?

Reply:
fixed